codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/EventStreamMarshallerGenerator.kt

```diff
@@ -122,34 +167,57 @@ class EventStreamMarshallerGenerator(
                 else -> throw IllegalStateException("unreachable")
             }
             addStringHeader(":content-type", "${ctx.contentType.dq()}.into()")
-            if (member.isOptional) {
-                rust(
-                    """
-                    if let Some(inner_payload) = inner.$memberName {
-                        inner_payload.${ctx.conversionFn}()
-                    } else {
-                        Vec::new()
-                    }
-                    """
-                )
-            } else {
-                rust("inner.$memberName.${ctx.conversionFn}()")
-            }
+            handleOptional(
+                optional,
+                inputExpr,
+                "inner_payload",
+                { input -> rust("$input.${ctx.conversionFn}()") },
+                { rust("Vec::new()") }
+            )
         } else {
             addStringHeader(":content-type", "${payloadContentType.dq()}.into()")
 
             val serializerFn = serializerGenerator.payloadSerializer(member)
-            rustTemplate(
-                """
-                        #{serializerFn}(&inner.$memberName)
+            handleOptional(
+                optional,
+                inputExpr,
+                "inner_payload",
+                { input ->
+                    rustTemplate(
+                        """
+                        #{serializerFn}(&$input)
                             .map_err(|err| #{Error}::Marshalling(format!("{}", err)))?
                         """,
-                "serializerFn" to serializerFn,
-                *codegenScope
+                        "serializerFn" to serializerFn,
+                        *codegenScope
+                    )
+                },
+                { rust("unimplemented!(\"TODO(EventStream): Figure out what to do when there's no payload\")") }
```

Comment:
this seems like the kind of runtime panic we should be _very_ cautious about since it's not that hard to trigger. Feels like a level of TODO a bit higher than the others we should be very careful not to launch with